### PR TITLE
Canonicalize collections in builder pattern (not finished yet)

### PR DESCRIPTION
**Pull request by yilongli**
for testing

### CODE REVIEW

File: src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java [hunk 1]
```diff
@@ -160,20 +154,39 @@ public boolean equals(Object object) {
...
+            return false;
+        }
     }
 
     @Override
     protected int computeHash() {
         int hashCode = 1;
-        hashCode = hashCode * Utils.HASH_PRIME + (frame == null ? 0 : frame.hashCode());
+        hashCode = hashCode * Utils.HASH_PRIME + baseTerms.hashCode();
         hashCode = hashCode * Utils.HASH_PRIME + cells.hashCode();
```
Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java:260]: Is it possible for two objects to test equal but have different hashcodes?

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java:260]: it shouldn't, right? I have changed `baseTerms` to be a set.

Comment (andreistefanescu) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java:260]: But then why do you not simply compare `frame`, the `baseTerms`, and `cells` in equals?

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java:260]: because when I added the `baseTerms` a while ago, I considered it as containing all variables of sort Bag, that is, including the frame

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/CellCollection.java:260]: Wait a minute, @andreistefanescu you should review pull request https://github.com/kframework/k/pull/915 instead. I created this pull request only for testing. So it contains several commits which belong to another pull request. Thanks.

**Comment by yilongli** [comment]
Jenkins: retest this please

**Comment by yilongli** [comment]
This has become a mess. I'm going to close this one and split the changes into two parts.
